Pythonscripting

HolgerHees opened the pull request:
Initial release with detailed description in marketplace

https://community.openhab.org/t/python3-scripting/161945

ccutrer commented:
FYI, I plan to use GraalPython with the MQTT Home Assistant binding so that I can run _actual_ Jinja, and not Jinjava which has had significant compatibility issues with more complex templates that devices can provide. We should probably avoid having multiple copies of the GraalPython bundles loaded, so I've started work on OSGi-ifying the Python specific bundles [here](https://github.com/openhab/openhab-osgiify/pull/58).  You can reference https://github.com/openhab/openhab-addons/pull/18053 for how to begin depending on the OSGi-ified versions of bundles.

florian-h05 commented:
@HolgerHees You requested my review - unfortunately I am unavailable for such a large review right now, but I should be able to have a look in March. Let me know if this is the final version now or not.

HolgerHees commented:
> @HolgerHees You requested my review - unfortunately I am unavailable for such a large review right now, but I should be able to have a look in March. Let me know if this is the final version now or not.

No rush, my goal is to merge the binding before the release of openhab 5 so that it can be released together. So we still have some time.

Currently it is the final version. At least there are no further feature requests or comments. And several people, including me, are already using it with openhab 4.3.2 as a production system since weeks without crashes or memory leaks. (except the know bugs related to truffle library loading, which still exists in openhab 4.3.x)

Maybe you know someone additionally for the review, to share the work a bit.

For now, my intention is to get early feedback about coding style, how to implement something better, something is missing or just finding bugs. Also I want to bring this project into the focus of much more people.

florian-h05 commented:
>No rush, my goal is to merge the binding before the release of openhab 5 so that it can be released together. So we still have some time.

:+1: 

> Maybe you know someone additionally for the review, to share the work a bit.

Hmm, for the Python scripts itself I am out, I usually go out of Python's way. For the add-on part, we need one of the add-on maintainers for the review and merge, but I can have a look at it, this should speed up review and last time a scripting add-on had to be reviewed, I was asked for help, so this might happen this time again (seems that all my work on JS Scripting makes me a scripting add-on expert ;-)).

> For now, my intention is to get early feedback about coding style, how to implement something better, something is missing or just finding bugs. 

There are some guidelines in the docs (https://next.openhab.org/docs/developer/guidelines.html) and there is a list what reviewers should look for, but you can check this list yourself to ease review (see https://github.com/openhab/openhab-addons/wiki/Review-Checklist).

One question though: Do you really want to put the Python scripts of the helper library into the add-on code? JS Scripting and JRuby Scripting have the helper library in a different repo and pull it in during build time (JS) or during runtime (JRuby). This allows you to develop the helper library without having to wait for the add-on maintainers, which I guess won't be big fans of reviewing Python. It would also allow the users to upgrade the helper library independently from openHAB itself.

HolgerHees commented:
Sounds good. In any case, I have some homework to do first :-)

I'm currently adapting the ScriptExtensionModuleProvider from the javascripting binding to suit my needs. It's already halfway running :-) and will make it possible to include all JSR223 objects via python imports instead of placing them in the global context by default.

Afterwards, I'll transfer the Python Helper Lib to a separate repo.

Thanks in advance for the helpful feedback.

HolgerHees commented:
python import wrapper is implemented and python code is moved to https://github.com/HolgerHees/openhab-python . This could be converted to a openhab repo later.

jlaur commented on bundles/org.openhab.automation.pythonscripting/src/main/java/org/openhab/automation/pythonscripting/internal/graal/GraalPythonScriptEngine.java:
```diff
@@ -0,0 +1,433 @@
+        Bindings bindings = this.getBindings(ScriptContext.ENGINE_SCOPE);
+        if (bindings instanceof GraalPythonBindings) {
+            ((GraalPythonBindings) bindings).updateEngineScriptContext(null);
+            // ((GraalPythonBindings) bindings).close();
```
Looks like WIP, but just to not forget.

HolgerHees replied:
changed

jlaur commented on bundles/org.openhab.automation.pythonscripting/src/main/java/org/openhab/automation/pythonscripting/internal/wrapper/ScriptExtensionModuleProvider.java:
```diff
@@ -0,0 +1,122 @@
+            } else if (name.startsWith(SCOPE_MODULE_PREFIX)) {
+                String[] segments = name.split("\\.");
+                if (name.equals(SCOPE_MODULE_PREFIX)) {
+                    Map<String, Object> _symbols = new HashMap<String, Object>(this.globals); // scriptExtensionAccessor.findDefaultPresets(engineIdentifier);
```
Left-over?

jlaur commented on bundles/org.openhab.automation.pythonscripting/src/main/java/org/openhab/automation/pythonscripting/internal/wrapper/ScriptExtensionModuleProvider.java:
```diff
@@ -0,0 +1,122 @@
+        this.globals.put(key, value);
+    }
+
+    // private Value toValue(Context ctx, Map<String, Object> map) {
+    // try {
+    // ctx.eval(Source.newBuilder( // convert to Map to Python Object
+    // GraalPythonScriptEngine.LANGUAGE_ID, //
+    // "import polyglot\n" //
+    // + "@polyglot.export_value\n" //
+    // + "def convert(map):\n" //
+    // + " obj = {}\n" //
+    // + " for prop in map.keySet():\n" //
+    // + " obj[prop] = map[prop]\n" //
+    // + " return obj\n" //
+    // + "",
+    // "<generated>").build());
+    // return ctx.getPolyglotBindings().getMember("convert").execute(map);
+    // } catch (IOException e) {
+    // throw new IllegalArgumentException("Failed to generate exports", e);
+    // }
+    // }
```
Left-over?

HolgerHees replied:
cleaned

jlaur commented on bundles/org.openhab.automation.pythonscripting/src/main/resources/OH-INF/addon/addon.xml:
```diff
@@ -0,0 +1,13 @@
+	<description>This adds a Python script engine.</description>
+	<connection>none</connection>
+
+	<service-id>org.openhab.pythonscripting</service-id>
```
```suggestion
	<service-id>org.openhab.automation.pythonscripting</service-id>
```

HolgerHees replied:
fixed

jlaur reviewed (commented):
Thanks for adding another automation, that's awesome! While you await a review and still are working on this, you could have a look also at compiler and checkstyle warnings. See target/code-analysis/report.html.

I have added a few quick minor comments, but only scratching the surface while peeking.

HolgerHees commented:
> Thanks for adding another automation, that's awesome! While you await a review and still are working on this, you could have a look also at compiler and checkstyle warnings. See target/code-analysis/report.html.
> 
> I have added a few quick minor comments, but only scratching the surface while peeking.

thanks for the valuable feedback. I fixed all places related to your commends. Most things are leftovers where I was not sure if it is really needed. Now, I checked these places again and refactored or cleaned it.

Also most of the hints from code analysis is fixed. Except 8 places with the recommendation to set "NonNullByDefault" fro classes. Most of theses classes extending other classes which makes it difficult.

HolgerHees commented on bundles/org.openhab.automation.pythonscripting/src/main/java/org/openhab/automation/pythonscripting/internal/PythonScriptEngine.java:
```diff
@@ -0,0 +1,457 @@
+    private final Logger logger = LoggerFactory.getLogger(PythonScriptEngine.class);
+
+    private static final String PYTHON_OPTION_EXECUTABLE = "python.Executable";
+    // private static final String PYTHON_OPTION_PYTHONHOME = "python.PythonHome";
```
done

HolgerHees commented on bundles/org.openhab.automation.pythonscripting/src/main/java/org/openhab/automation/pythonscripting/internal/PythonScriptEngine.java:
```diff
@@ -0,0 +1,457 @@
+            .targetTypeMapping(Value.class, Item.class, v -> v.hasMember("raw_item"),
+                    v -> v.getMember("raw_item").as(Item.class), HostAccess.TargetMappingPrecedence.LOW)
+
+            // Translate python quantity to org.openhab.core.library.types.QuantityType
+            // .targetTypeMapping(Value.class, QuantityType.class, v -> v.hasMember("rawQtyType"),
+            // v -> v.getMember("rawQtyType").as(QuantityType.class), HostAccess.TargetMappingPrecedence.LOW)
+
```
cleaned

HolgerHees commented on bundles/org.openhab.automation.pythonscripting/src/main/java/org/openhab/automation/pythonscripting/internal/PythonScriptEngine.java:
```diff
@@ -0,0 +1,457 @@
+                            }
+                        }).build()) //
+                .allowHostAccess(HOST_ACCESS) //
+                // .allowHostClassLoading(true) //
```
refactored

HolgerHees commented on bundles/org.openhab.automation.pythonscripting/src/main/java/org/openhab/automation/pythonscripting/internal/PythonScriptEngine.java:
```diff
@@ -0,0 +1,457 @@
+                // to discover packages
+                .option(PYTHON_OPTION_EXECUTABLE,
+                        PythonScriptEngineFactory.PYTHON_DEFAULT_PATH.resolve("bin").resolve("python").toString())
+                // Set the python home to be read from the embedded resources
+                // .option(PYTHON_OPTION_PYTHONHOME, PYTHON_DEFAULT_PATH.toString()) //
```
refactored

HolgerHees commented on bundles/org.openhab.automation.pythonscripting/src/main/java/org/openhab/automation/pythonscripting/internal/PythonScriptEngine.java:
```diff
@@ -0,0 +1,457 @@
+    }
+
+    private static Set<String> transformGraalWrapperSet(Value value) {
+        // Value raw_value = value.invokeMember("getWrappedSetValues");
```
cleaned

HolgerHees commented on bundles/org.openhab.automation.pythonscripting/src/main/java/org/openhab/automation/pythonscripting/internal/PythonScriptEngine.java:
```diff
@@ -0,0 +1,457 @@
+
+    private static class LogOutputStream extends OutputStream {
+        private static final int DEFAULT_BUFFER_LENGTH = 2048;
+        private static final String LINE_SEPERATOR = System.getProperty("line.separator");
```
changed

HolgerHees commented on bundles/org.openhab.automation.pythonscripting/src/main/java/org/openhab/automation/pythonscripting/internal/PythonScriptEngine.java:
```diff
@@ -0,0 +1,457 @@
+    private static class LogOutputStream extends OutputStream {
+        private static final int DEFAULT_BUFFER_LENGTH = 2048;
+        private static final String LINE_SEPERATOR = System.getProperty("line.separator");
+        private static final int LINE_SEPERATOR_SIZE = LINE_SEPERATOR.length();
```
changed

HolgerHees commented on bundles/org.openhab.automation.pythonscripting/src/main/java/org/openhab/automation/pythonscripting/internal/PythonScriptEngineFactory.java:
```diff
@@ -0,0 +1,243 @@
+ * @author Holger Hees - initial contribution
+ * @author Jeff James - initial contribution
+ */
+@Component(service = ScriptEngineFactory.class, configurationPid = "org.openhab.pythonscripting", property = Constants.SERVICE_PID
+        + "=org.openhab.pythonscripting")
```
fixed

HolgerHees commented on bundles/org.openhab.automation.pythonscripting/src/main/java/org/openhab/automation/pythonscripting/internal/graal/GraalPythonBindings.java:
```diff
@@ -0,0 +1,140 @@
+    }
+
+    private static void checkKey(String key) {
+        Objects.requireNonNull(key, "key can not be null");
```
refactored

HolgerHees commented on bundles/org.openhab.automation.pythonscripting/src/main/java/org/openhab/automation/pythonscripting/internal/graal/GraalPythonScriptEngine.java:
```diff
@@ -0,0 +1,433 @@
+                    .allowExperimentalOptions(true) //
+                    .allowAllAccess(true) //
+                    .allowHostAccess(HostAccess.ALL) //
+                    // TODO .allowIO(IOAccess.newBuilder().allowHostSocketAccess(true).fileSystem(null).build()) //
```
cleaned

HolgerHees commented on bundles/org.openhab.automation.pythonscripting/src/main/java/org/openhab/automation/pythonscripting/internal/graal/GraalPythonScriptEngine.java:
```diff
@@ -0,0 +1,433 @@
+     */
+    @Override
+    public void close() {
+        logger.info("GraalPythonScriptEngine closed");
```
fixed

jlaur commented on bundles/org.openhab.automation.pythonscripting/src/main/java/org/openhab/automation/pythonscripting/internal/graal/GraalPythonBindings.java:
```diff
@@ -0,0 +1,127 @@
+
+import javax.script.ScriptContext;
+import javax.script.ScriptEngine;
+import javax.validation.constraints.NotNull;
```
I'm not sure if there is a specific reason for using that here, but generally we use the null annotations from `org.eclipse.jdt.annotation`, e.g. `org.eclipse.jdt.annotation.NonNull`.

HolgerHees replied:
was introduced by a mistake. But now I checked again and it was no necessary anymore. So it is cleaned now.

mbronk commented on bundles/org.openhab.automation.pythonscripting/pom.xml:
```diff
@@ -0,0 +1,262 @@
+            </goals>
+            <phase>generate-resources</phase>
+            <configuration>
+              <connectionUrl>scm:git:https://github.com/HolgerHees/openhab-python</connectionUrl>
```
It would be great to move this repo under the OpenHAB [GH org](https://github.com/openhab), similar to how JRuby helpers are present [here](https://github.com/openhab/openhab-jruby/).
This way not only would the repo attract more maintainers and gain governance, but it'll also allow to version these libs together with OH releases (which I think we should consider to stabilize these interfaces).
**WDYT**?

----
_Side note_: I'm no maintainer nor do I have any authoritative say in the matter, so feel free to disregard this suggestion entirely. Just stumbled across this PR _(awesome work, BTW 🙏 )_ and considered it would be more "secure" if the build did not bundle "external" resources.

HolgerHees replied:
Hi @mbronk ,

That's the plan. I want to move it after the review, but fore the final release and together with openHAB 5. But for the starting phase, I have not find a better place :-)

HolgerHees commented:
Hi @jlaur 

There is a follow-up pull request where all your feedback has been considered and implemented.  And it is already approved by @florian-h05 .

https://github.com/openhab/openhab-addons/pull/18499

I closed this pull request because various main branch merges caused problems with the rebase. I now realize that it could have been handled differently.
